Quantum/symplecticKickedRotor.py

```diff
@@ -6,51 +6,86 @@
 import matplotlib.pyplot as plt
 
+import time
+
 class SpinKickedRotor():
     
-    def __init__(self,kb = 2.89):   
+    def __init__(self, mu, epsilon, kb = 2.89):   
         self.I  = complex(0,1)
         self.kb = kb
+        self.mu = mu
+        self.epsilon = epsilon
         
-    def Ukick(self, x, Psi, K):
+    def Ukick(self, x, K):
         """Input : p -> array, impulsions
                    Psi -> 2d array, initial state
                    K -> float, kick's strength
            Output : array, returns the state after a Kick"""
         
-        Uk = np.array([[np.cos(x) + np.sin(x),np.sin(2*x)],[np.sin(2*x),np.cos(x) - np.sin(x)]])
-        Uk = np.exp(-(self.I*K/self.kb)*Uk)
+        L = len(x)
+        
+        x_diag = np.diag(x)
         
-        f1 = fft.fftshift(fft.fft(fft.ifftshift(Psi[0])))
-        f2 = fft.fftshift(fft.fft(fft.ifftshift(Psi[1])))
+        # fk = fft.fftshift(fft.fft(fft.ifftshift(Psi)))
         
-        res_kick = np.array([Uk[0,0]*f1 + Uk[0,1]*f2, Uk[1,0]*f1 + Uk[1,1]*f2])
+        Uk = np.zeros((2*L,2*L))
         
-        return res_kick
+        U1 = np.cos(x_diag) + self.epsilon * np.sin(x_diag)
+        U2 = 1/2 * self.mu * np.sin(2*x_diag)
+        U3 = 1/2 * self.mu *np.sin(2*x_diag)
+        U4 = np.cos(x_diag) - self.epsilon * np.sin(x_diag)
+        
+        Uk[:L,:L]       = U1
+        Uk[:L,L:2*L]    = U2
+        Uk[L:2*L,:L]    = U3
+        Uk[L:2*L,L:2*L] = U4
+        
+        Uk = expm(-(self.I*K/self.kb)*Uk)
+                
+        return Uk
         
-    def Uprop(self, x, Psi, b=0):
+    def Uprop(self, p,b=0):
         """Input : p -> array, impulsions
                    Psi -> 2d array, initial state
                    b -> float, pseudo-impulsion
            Output : array, returns the state after a Propagation"""
            
-        zeros = np.zeros(len(x))
+        L = len(p)
         
-        Up = np.array([[(x + self.kb*b)**2,zeros],[zeros,(x + self.kb*b)**2]],dtype=complex)
-        Up = np.exp(-(self.I*self.kb/2) * Up)
+        p_diag = np.diag(p)
+        zeros = np.zeros((L,L))
         
-        f1 = fft.ifftshift(fft.ifft(fft.fftshift(Psi[0])))
-        f2 = fft.ifftshift(fft.ifft(fft.fftshift(Psi[1])))
+        #fp = fft.ifftshift(fft.ifft(fft.fftshift(Psi)))
         
-        res_prop = np.array([Up[0,0]*f1 + Up[0,1]*f2, Up[1,0]*f1 + Up[1,1]*f2])
-        return res_prop    
+        Up = np.zeros((2*L,2*L))
+        
+        U1 = ((p_diag + b)**2)
+        U2 = zeros
+        U3 = zeros
+        U4 = ((p_diag + b)**2)
+        
+        Up[:L,:L]       = U1
+        Up[:L,L:2*L]    = U2
+        Up[L:2*L,:L]    = U3
+        Up[L:2*L,L:2*L] = U4
+
+        Up = expm(-(self.I*self.kb/2) * Up)
+        
+        # f1 = fft.ifftshift(fft.ifft(fft.fftshift(Psi[0])))
+        # f2 = fft.ifftshift(fft.ifft(fft.fftshift(Psi[1])))
+        
+        # res_prop = np.matmul(Up,fp)
+        return Up    
     
     def loop(self, x, p, Psi, K, b, n):
```

Comment:
The loop function doesn't have any point since the next method doesn't call it anymore. It will be remove in a next version eventually.